rina-tools/src/rlite/rinaperf.c

```diff
@@ -560,24 +791,160 @@ static void *
+        }
+        goto out;
+    }
 
-    if (w->test_config.ty >= sizeof(descs)) {
+    cfg.opcode  = le32toh(cfg.opcode);
+    cfg.ticket  = le32toh(cfg.ticket);
+    cfg.cnt     = le64toh(cfg.cnt);
+    cfg.size    = le32toh(cfg.size);
+
+    if (cfg.opcode >= RP_OPCODE_STOP) {
+        printf("Invalid test configuration: test type %u is invalid\n", cfg.opcode);
         goto out;
     }
 
-    w->fn(w);
+    if (cfg.opcode == RP_OPCODE_DATAFLOW) {
+        /* This is a data flow. We need to pass the file descriptor to the worker
+         * associated to the ticket, and notify it. */
+
+        pthread_mutex_lock(&rp->ticket_lock);
+        if (cfg.ticket >= RP_MAX_WORKERS || !rp->ticket_table[cfg.ticket]) {
+            printf("Invalid ticket request: ticket %u is invalid\n", cfg.ticket);
+        } else {
+            struct worker *tw = rp->ticket_table[cfg.ticket];
+
+            rp->ticket_table[cfg.ticket] = NULL;
+#if 0
+            printf("TicketTable: data flow for ticket %u\n", cfg.ticket);
+#endif
+            tw->dfd = w->cfd;
+            w->cfd = -1;
+            pthread_cond_signal(&tw->data_flow_ready);
+        }
+        pthread_mutex_unlock(&rp->ticket_lock);
+    } else {
+        /* This is a control flow. */
+        if (cfg.size < sizeof(uint16_t)) {
+            printf("Invalid test configuration: size %u is invalid\n", cfg.size);
+            goto out;
+        }
+
+        /* Allocate a ticket for the client. */
+        {
+            pthread_mutex_lock(&rp->ticket_lock);
+            for (ticket = 0; ticket < RP_MAX_WORKERS; ticket ++) {
+                if (rp->ticket_table[ticket] == NULL) {
+                    rp->ticket_table[ticket] = w;
+                    break;
+                }
+            }
+            pthread_cond_init(&w->data_flow_ready, NULL);
+#if 0
+            printf("TicketTable: allocated ticket %u\n", ticket);
+#endif
+            pthread_mutex_unlock(&rp->ticket_lock);
+            assert(ticket < RP_MAX_WORKERS);
+        }
+
+        /* Send ticket back to the client. */
+        tmsg.ticket = htole32(ticket);
+        ret = write(w->cfd, &tmsg, sizeof(tmsg));
+        if (ret != sizeof(tmsg)) {
+            if (ret < 0) {
+                perror("write(ticket)");
+            } else {
+                printf("Error writing ticket: wrong length %d "
+                       "(should be %lu)\n", ret, (unsigned long int)sizeof(tmsg));
+            }
+            goto out;
+        }
+
+        if (rp->verbose) {
+            printf("Configuring test type %u, SDU count %lu, SDU size %u, "
+                    "ticket %u\n",
+                   cfg.opcode, cfg.cnt, cfg.size, ticket);
+        }
+
+        /* Wait for the client to allocate a data flow and come back to us. */
+        pthread_mutex_lock(&rp->ticket_lock);
+        clock_gettime(CLOCK_REALTIME, &to);
+        to.tv_sec += 5;
+        ret = pthread_cond_timedwait(&w->data_flow_ready,
```

Comment:
It's because of the threading implementation in the kernel. Even if you didn't call pthread_cond_signal, it can still wake up. So you need to do a while(cond == false) cond_wait();